File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int bfs(int node, int n, vector<vector<int>> &computer, vector<int> &visited) {
    queue<int> q;
    
    q.push(node);
    visited[node] = 1;
    
    while (!q.empty()) {
        
        int currentNode = q.front();
        q.pop();
        
        for (int i = 0; i < n; i++) {
            if (i == currentNode)
                continue;

            if (computer[currentNode][i] == 1 && visited[i] == 0) {
                q.push(i);
                visited[i] = 1;
            }
        }
    }
    return 1;
    
}

int solution(int n, vector<vector<int>> computers) {
    
    if (n == 1)
        return 1;
    
    int answer = 0;
    
    vector<int> visited (n, 0);
    
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (i == j)
                continue;
            
            if (visited[i] != 1) {
                answer += bfs(i, n, computers, visited);
            }
        }
    }
    return answer;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.py
visited = list()


def bfs(current, n, computers):
    q = list()
    q.append(current)
    visited[current] = 0
    
    while q:
        node = q[0]
        del q[0]
        
        for idx, connected in enumerate(computers[node]):
            if connected == 1 and visited[idx] == 0:
                q.append(idx)
                visited[idx] = 1
                
    return 1;
            
    

def solution(n, computers):
    answer = 0
    
    global visited
    visited = [0 for i in range(n)]
    
    for i in range(len(computers)):
        if visited[i] == 0:
            answer += bfs(i, n, computers)

    return answer